pkg/rbac/parser.go

```diff
@@ -63,83 +61,93 @@ type Rule struct {
-func (keys ruleKeys) Swap(i, j int)      { keys[i], keys[j] = keys[j], keys[i] }
-func (keys ruleKeys) Less(i, j int) bool { return keys[i].String() < keys[j].String() }
-
-// key normalizes the Rule and returns a ruleKey object.
-func (r *Rule) key() ruleKey {
-	r.normalize()
-	return ruleKey{
-		Groups:        strings.Join(r.Groups, "&"),
-		Resources:     strings.Join(r.Resources, "&"),
-		ResourceNames: strings.Join(r.ResourceNames, "&"),
-		URLs:          strings.Join(r.URLs, "&"),
+func (r *Rule) Normalize() *NormalizedRule {
+	result := &NormalizedRule{
+		Groups:        newSet(r.Groups),
+		Resources:     newSet(r.Resources),
+		ResourceNames: newSet(r.ResourceNames),
+		Verbs:         newSet(r.Verbs),
+		URLs:          newSet(r.URLs),
+		Namespace:     r.Namespace,
+	}
+
+	// fix the group names, since letting people type "core" is nice
+	if _, ok := result.Groups["core"]; ok {
+		delete(result.Groups, "core")
+		result.Groups[""] = struct{}{}
 	}
+
+	result.GenerateComparisonKey()
+
+	return result
 }
 
-// addVerbs adds new verbs into a Rule.
-// The duplicates in `r.Verbs` will be removed, and then `r.Verbs` will be sorted.
-func (r *Rule) addVerbs(verbs []string) {
-	r.Verbs = removeDupAndSort(append(r.Verbs, verbs...))
+type NormalizedRule struct {
+	// if two different rules have the same comparison key then they can have their verbs merged
+	// this key should comprise all the fields below except Verbs and Namespace (since rules
+	// are partitioned by namespace before merging)
+	ComparisonKey string
+
+	Namespace     string
+	Groups        sets.String
+	Resources     sets.String
+	ResourceNames sets.String
+	URLs          sets.String
+
+	Verbs sets.String
 }
 
-// normalize removes duplicates from each field of a Rule, and sorts each field.
-func (r *Rule) normalize() {
-	r.Groups = removeDupAndSort(r.Groups)
-	r.Resources = removeDupAndSort(r.Resources)
-	r.ResourceNames = removeDupAndSort(r.ResourceNames)
-	r.Verbs = removeDupAndSort(r.Verbs)
-	r.URLs = removeDupAndSort(r.URLs)
+// GenerateComparisonKey generates the ComparisonKey
+func (nr *NormalizedRule) GenerateComparisonKey() {
+	nr.ComparisonKey = strings.Join(
+		[]string{
+			strings.Join(setToSorted(nr.Groups), "&"),
+			strings.Join(setToSorted(nr.Resources), "&"),
+			strings.Join(setToSorted(nr.ResourceNames), "&"),
+			strings.Join(setToSorted(nr.URLs), "&"),
+		},
+		" + ")
 }
 
-// removeDupAndSort removes duplicates in strs, sorts the items, and returns a
-// new slice of strings.
-func removeDupAndSort(strs []string) []string {
-	set := make(map[string]bool)
-	for _, str := range strs {
-		if _, ok := set[str]; !ok {
-			set[str] = true
-		}
+func setToSorted(set sets.String) []string {
+	if set == nil { // preserve nils
+		return nil
 	}
 
-	var result []string
-	for str := range set {
-		result = append(result, str)
-	}
+	result := set.UnsortedList()
 	sort.Strings(result)
 	return result
 }
 
+// Subsumes indicates if one rule entirely determines another,
```

Comment:
This might be due to the fact that I am not a native English speaker, but I find a name like `IsSuperset` way more obvious.